File: lena-config/src/main/java/org/riversoforion/lena/config/Name.java
/*
 * Copyright (c) 2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config;

import java.text.Normalizer.Form;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.StringJoiner;
import java.util.regex.Pattern;
import java.util.stream.Stream;

import static java.text.Normalizer.normalize;
import static java.util.Arrays.asList;

public class Name implements CharSequence, Comparable<Name> {

    public static final String SEPARATOR = "/";
    private static final Pattern NON_ASCII = Pattern.compile("[^\\p{ASCII}]");
    private static final Pattern NON_LETTERS_OR_NUMBERS = Pattern.compile("[^a-z0-9]+");

    private final List<String> rawSegments;
    private final String fullName;

    private Name(List<String> segments, boolean allowEmpty) {

        this.rawSegments = segments == null ? List.of() : segments.stream().filter(Name::notEmpty).flatMap(Name::splitSegment).toList();
        if (rawSegments.isEmpty() && !allowEmpty) {
            throw new IllegalArgumentException("Name must contain at least one segment");
        }
        StringJoiner joiner = new StringJoiner(SEPARATOR);
        this.rawSegments.stream().map(Name::normalizeSegment).filter(Name::notEmpty).forEach(joiner::add);
        this.fullName = joiner.toString();
    }

    static Stream<String> splitSegment(String segment) {

        return Stream.of(segment.split(SEPARATOR)).filter(Name::notEmpty);
    }

    static String normalizeSegment(String segment) {

        if (segment == null || segment.isBlank()) {
            return "";
        }
        String partiallyNormalized = NON_ASCII.matcher(normalize(segment.trim().toLowerCase(Locale.ROOT), Form.NFD)).replaceAll("");
        return NON_LETTERS_OR_NUMBERS.matcher(partiallyNormalized).replaceAll("-");
    }

    private static boolean notEmpty(String segment) {

        return segment != null && !segment.isBlank();
    }

    // Static factories

    public static Name of(String... segments) {

        if (segments == null || segments.length == 0) {
            throw new IllegalArgumentException("Name must contain at least one segment");
        }
        return new Name(asList(segments), false);
    }

    public static Name of(List<String> segments) {

        if (segments == null || segments.isEmpty()) {
            throw new IllegalArgumentException("Name must contain at least one segment");
        }
        return new Name(segments, false);
    }

    static Name forNamespace(List<String> segments) {

        return new Name(segments, true);
    }

    // Operations

    public List<String> segments() {

        return List.copyOf(rawSegments);
    }

    // Object

    @Override
    public String toString() {

        return fullName;
    }

    @Override
    public boolean equals(Object o) {

        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Name namespace = (Name) o;
        return Objects.equals(fullName, namespace.fullName);
    }

    @Override
    public int hashCode() {

        return Objects.hashCode(fullName);
    }

    // CharSequence

    @Override
    public int length() {

        return fullName.length();
    }

    @Override
    public char charAt(int index) {

        return fullName.charAt(index);
    }

    @Override
    public CharSequence subSequence(int start, int end) {

        return fullName.subSequence(start, end);
    }

    // Comparable

    @Override
    public int compareTo(Name o) {

        if (o == null) {
            return 1;
        }
        return fullName.compareTo(o.fullName);
    }
}

File: lena-config/src/commonMain/kotlin/org/riversoforion/lena/config/Name.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config

import org.riversoforion.lena.config.internal.normalizeForName

/**
 * An immutable, normalized hierarchical configuration property name.
 *
 * Segments are joined with [SEPARATOR] (`/`) and normalized: trimmed, lowercased,
 * non-ASCII characters removed (via Unicode NFD decomposition on JVM; best-effort on native),
 * and runs of non-alphanumeric characters replaced by `-`.
 *
 * Examples:
 * - `Name.of("service", "url")` → `"service/url"`
 * - `Name.of("NET", "read", "timeout")` → `"net/read/timeout"`
 */
public class Name private constructor(
    private val rawSegments: List<String>,
    private val fullName: String,
) : Comparable<Name> {

    public companion object {
        public const val SEPARATOR: String = "/"

        private val NON_ALPHANUMERIC: Regex = Regex("[^a-z0-9]+")

        public fun of(vararg segments: String): Name {
            require(segments.isNotEmpty()) { "Name must contain at least one segment" }
            return fromSegments(segments.toList(), allowEmpty = false)
        }

        public fun of(segments: List<String>): Name {
            require(segments.isNotEmpty()) { "Name must contain at least one segment" }
            return fromSegments(segments, allowEmpty = false)
        }

        internal fun forNamespace(segments: List<String>): Name =
            fromSegments(segments, allowEmpty = true)

        internal fun splitSegment(segment: String): List<String> =
            segment.split(SEPARATOR).filter { it.isNotBlank() }

        internal fun normalizeSegment(segment: String): String {
            if (segment.isBlank()) return ""
            val decomposed = normalizeForName(segment.trim())
            return NON_ALPHANUMERIC.replace(decomposed.lowercase(), "-")
        }

        private fun fromSegments(segments: List<String>, allowEmpty: Boolean): Name {
            val raw = segments
                .filter { it.isNotBlank() }
                .flatMap { splitSegment(it) }
            require(allowEmpty || raw.isNotEmpty()) { "Name must contain at least one segment" }
            val full = raw.mapNotNull { normalizeSegment(it).ifEmpty { null } }
                .joinToString(SEPARATOR)
            return Name(raw, full)
        }
    }

    public fun segments(): List<String> = rawSegments.toList()

    override fun toString(): String = fullName

    override fun equals(other: Any?): Boolean {
        if (other !is Name) return false
        return fullName == other.fullName
    }

    override fun hashCode(): Int = fullName.hashCode()

    override fun compareTo(other: Name): Int = fullName.compareTo(other.fullName)
}
